File: observer/Teacher.java
package observer;

import java.util.ArrayList;

public class Teacher implements Subject {
    private String _phone;

    public String get_phone() {
        return _phone;
    }

    public void set_phone(String _phone) {
        this._phone = _phone;
        notifyObserver();
    }

    public Teacher() {
        stuList = new ArrayList<>();
    }

    private ArrayList stuList;

    @Override
    public void registerObserver(Observer o) {
        stuList.add(o);
    }

    @Override
    public void removeObserver(Observer o) {
        stuList.remove(o);
    }

    @Override
    public void notifyObserver() {
        for (Object o : stuList) {
            ((Observer)o).update(get_phone());
        }

    }
}

File: Ts/observer/Teacher.ts
class Teacher implements Subject {
  private _phone: String;
  private stuList: Array<any>;
  constructor() {
    this.stuList = new Array<Observer>();
  }
  public get_phone(): String {
    return this._phone;
  }
  public set_phone(phone: String): void {
    this._phone = phone;
    this.notifyObserver();
  }
  registerObserver(o: Observer) {
    this.stuList.push(o);
  }
  removeObserver(o: Observer) {
    this.stuList = this.stuList.filter(item => item !== o);
  }
  notifyObserver() {
    this.stuList.forEach(item => {
      item.update(this.get_phone());
    })
  }

  
}